golangci-lint: suppress one issue, demote others to "hints"

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup


#### What this PR does / why we need it:

The voting in https://github.com/kubernetes/kubernetes/issues/117288 led to one check that got rejected ("ifElseChain: rewrite if-else to switch statement") and several that are "nice to know".

golangci-lint's support for issue "severity" is too limited to identify "nice to know" issues in the output (filtering is only by linter without considering the issue text; not part of text output). Therefore a third configuration gets added which emits all issues (must fix and nits). The intention is to use the "strict" configuration in pull-kubernetes-verify and the "hints" configuration in a new non-blocking pull-kubernetes-linter-hints.

That way, "must fix" issues will block merging while issues that may be useful will show up in a failed optional job. However, that job then also contains "must fix" issues, partly because filtering out those would make the configuration a lot larger and is likely to be unreliably (all "must fix" issues would need to be identified and listed), partly because it may be useful to have all issues in one place.

#### Which issue(s) this PR fixes:
Fixes #117288

#### Special notes for your reviewer:

The previous approach of manually keeping two configs in sync with special comments didn't scale to three configs. Now a single golangci.yaml.in with text/template constructs contains the source for all three configs. A new simple CLI frontend for text/template (cmd/gotemplate) is used by hack/update-golangci-lint-config.sh to generate the three flavors.

Verification uses the same approach as hack/verify-codegen.sh (ensure working tree is clean, update, check for a diff). Because the logic is the same and only the update command is different, the common logic gets moved into hack/lib/verify-any.sh.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
golangci-lint: suppress one issue, demote others to "hints"

The voting in https://github.com/kubernetes/kubernetes/issues/117288 led to
one check that got rejected ("ifElseChain: rewrite if-else to switch
statement") and several that are "nice to know".

golangci-lint's support for issue "severity" is too limited to identify "nice
to know" issues in the output (filtering is only by linter without considering
the issue text; not part of text output). Therefore a third configuration gets
added which emits all issues (must fix and nits). The intention is to use
the "strict" configuration in pull-kubernetes-verify and the "hints"
configuration in a new non-blocking pull-kubernetes-linter-hints.

That way, "must fix" issues will block merging while issues that may be useful
will show up in a failed optional job. However, that job then also contains
"must fix" issues, partly because filtering out those would make the
configuration a lot larger and is likely to be unreliably (all "must fix"
issues would need to be identified and listed), partly because it may be useful
to have all issues in one place.

The previous approach of manually keeping two configs in sync with special
comments didn't scale to three configs. Now a single golangci.yaml.in with
text/template constructs contains the source for all three configs. A new
simple CLI frontend for text/template (cmd/gotemplate) is used by
hack/update-golangci-lint-config.sh to generate the three flavors.

pohly commented:
/test pull-kubernetes-verify-strict-lint

dims commented:
/assign @BenTheElder

thockin commented on hack/lib/verify-any.sh:
```diff
@@ -0,0 +1,51 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script verifies whether generated files are up-to-date. The
```
Example here, please?

pohly replied:
Added.

thockin commented on hack/lib/verify-any.sh:
```diff
@@ -0,0 +1,51 @@
+# This sets up the environment, like GOCACHE, which keeps the worktree cleaner.
+kube::golang::setup_env
+
+_tmpdir="$(kube::realpath "$(mktemp -d -t "$(basename "$0").XXXXXX")")"
```
What is $0 here - is it "verify-any.sh"?  That's not super useful.  Maybe you should check that `$1` is not empty and use that here instead?

pohly replied:
In the new function, it is now `"verify-generated-$(basename "$1" | sed -e 's/[^a-zA-Z][^a-zA-Z]*/-/g').XXXXXX")"`.

`$1` cannot be empty, it must be a function or command that is to be executed.

thockin commented on hack/lib/verify-any.sh:
I don't love the name of this file (or that it is a file, really).  We have a ton of `verify*` files, this is not like them.  Since this is really a utility, why not turn it into function (`run_verify` ?) in hack/lib and call it from places like codegen (and mocks, and internal-modules, and licenses, and yamlfmt...)

That could be a commit of its own, just removing that duplication from those verify-* scripts.

pohly replied:
Done. Please take another look.

thockin commented on hack/update-golangci-lint-config.sh:
```diff
@@ -0,0 +1,50 @@
+
+# Inform about differences.
+echo
+diff -c hack/golangci.yaml hack/golangci-strict.yaml || true
```
If we are going to diff, can we use `-u` ?

pohly replied:
Changed. Here's what an invocation now looks like:

```console
$ hack/update-golangci-lint-config.sh 
+++ [0621 08:45:09] Setting GOMAXPROCS: 36
Generating hack/golangci.yaml from hack/golangci.yaml.in with ./cmd/gotemplate Base=1
Generating hack/golangci-strict.yaml from hack/golangci.yaml.in with ./cmd/gotemplate Strict=1
Generating hack/golangci-hints.yaml from hack/golangci.yaml.in with ./cmd/gotemplate Hints=1

--- hack/golangci.yaml	2023-06-21 08:45:10.222186135 +0200
+++ hack/golangci-strict.yaml	2023-06-21 08:45:10.478185292 +0200
@@ -30,11 +30,6 @@
     - path: conversion\.go
       linters:
         - ineffassign
-    # TODO(oscr) Remove these excluded directories and fix findings. Due to large amount of findings in different components
-    # with different owners it's hard to fix everything in a single pr. This will therefore be done in multiple prs.
-    - path: (pkg/volume/*|test/*|azure/*|pkg/cmd/wait*|request/bearertoken/*|metrics/*|filters/*)
-      linters:
-        - gocritic
 
     # The Kubernetes naming convention for conversion functions uses underscores
     # and intentionally deviates from normal Go conventions to make those function
@@ -88,7 +83,7 @@
       text: "ST1023: should omit type .* from declaration; it will be inferred from the right-hand side"
 
 linters:
-  disable-all: true
+  disable-all: false
   enable: # please keep this alphabetized
     - ginkgolinter
     - gocritic
@@ -98,6 +93,9 @@
     - staticcheck
     - stylecheck
     - unused
+  disable:
+    # https://github.com/kubernetes/kubernetes/issues/117288#issuecomment-1507008359
+    - errcheck
 
 linters-settings: # please keep this alphabetized
   custom:
@@ -106,15 +104,6 @@
       path: ../_output/local/bin/logcheck.so
       description: structured logging checker
       original-url: k8s.io/logtools/logcheck
-  gocritic:
-    enabled-checks:
-      - equalFold
-      - boolExprSimplify
   staticcheck:
     checks:
       - "all"
-      - "-SA1019"  # TODO(fix) Using a deprecated function, variable, constant or field
-      - "-SA2002"  # TODO(fix) Called testing.T.FailNow or SkipNow in a goroutine, which isn’t allowed
-  stylecheck:
-    checks:
-      - "ST1019"   # Importing the same package multiple times

--- hack/golangci-strict.yaml	2023-06-21 08:45:10.478185292 +0200
+++ hack/golangci-hints.yaml	2023-06-21 08:45:10.734184449 +0200
@@ -46,42 +46,6 @@
         - gocritic
       text: "ifElseChain: rewrite if-else to switch statement"
 
-    # The following issues were deemed "might be worth fixing, needs to be
-    # decided on a case-by-case basis".  This was initially decided by a
-    # majority of the developers who voted in
-    # https://github.com/kubernetes/kubernetes/issues/117288 and may evolve
-    # over time.
-
-    # https://github.com/kubernetes/kubernetes/issues/117288#issuecomment-1507008918
-    - linters:
-        - gocritic
-      text: "assignOp:"
-
-    # https://github.com/kubernetes/kubernetes/issues/117288#issuecomment-1507016854
-    - linters:
-        - gosimple
-      text: "S1002: should omit comparison to bool constant"
-
-    # https://github.com/kubernetes/kubernetes/issues/117288#issuecomment-1507023980
-    - linters:
-        - gosimple
-      text: "S1016: should convert opts .* instead of using struct literal"
-
-    # https://github.com/kubernetes/kubernetes/issues/117288#issuecomment-1507026758
-    - linters:
-        - gosimple
-      text: "S1033: unnecessary guard around call to delete"
-
-    # https://github.com/kubernetes/kubernetes/issues/117288#issuecomment-1507030071
-    - linters:
-        - stylecheck
-      text: "ST1012: error var .* should have name of the form ErrFoo"
-
-    # https://github.com/kubernetes/kubernetes/issues/117288#issuecomment-1507031224
-    - linters:
-        - stylecheck
-      text: "ST1023: should omit type .* from declaration; it will be inferred from the right-hand side"
-
 linters:
   disable-all: false
   enable: # please keep this alphabetized
@@ -93,9 +57,6 @@
     - staticcheck
     - stylecheck
     - unused
-  disable:
-    # https://github.com/kubernetes/kubernetes/issues/117288#issuecomment-1507008359
-    - errcheck
 
 linters-settings: # please keep this alphabetized
   custom:
```

Is that perhaps too verbose?

pohly replied:
Diff is removed now.

thockin reviewed (commented):
Having a whole new tool to run gotemplates, which are pretty universally hated, seems unfortunate.  Do we have alternatives?  We probably don't want to demand that people have jinja install (or something), right?  Can we do this with sed or awk?

pohly commented:
> Can we do this with sed or awk?

I doubt that it would be nice. How about a `golangci-lint.in.sh` shell script that is one giant
```
cat <<EOF
# golangci-lint is used in Kubernetes with different configurations that
# enable an increasing amount of checks:
# - golangci.yaml is the most permissive configuration. All existing code
#   passed.
...
$(if $base; then cat <<EOF2
# some code only for the base config
EOF2))
...
EOF
```

Would that be better?

thockin commented:
I don't think multiple embedded heredocs are better, no :)

How about we run it through `cpp`

```
# golangci-lint is used in Kubernetes with different configurations that
# enable an increasing amount of checks:
# - golangci.yaml is the most permissive configuration. All existing code
#   passed.
...
#ifdef BASE
# some code only for the base config
#endif
...
```

I am kidding but only barely.

pohly commented:
> I don't think multiple embedded heredocs are better, no :)

They could get avoided by embedding calls to functions which then contain those heredocs, but that would break the flow of the config.

> How about we run it through cpp

We can't assume that cpp is in the `PATH`, though, can we?

pohly commented:
`cpp` is not capable enough: golangci-lint 1.16.x will support embedding the logcheck.conf inside the golangci-lint.yaml. But having that as a separate file is useful, so I want to include it in the template. That depends on indenting it, which `cpp` cannot do.

pohly committed:
hack: move common "verify generated" shell code into function

Several verify scripts used the same pattern of "check for clean working tree,
generated files, check for diffs". The code for that is now in
kube::verify::generated, defined in hack/lib/verify-generated.sh, and those
scripts just source that.

pohly commented:
/sig testing

thockin reviewed (commented):
Comments on first commit.  I noticed there's one more use of worktree in hack that you didn't capture, but it's not a "verify + update" pair.

We can do this as a followup (or not) but something like:

```
function run_in_worktree() {
    wtpfx="$1"
    shift

    (  # new sub-shell
        _tmpdir="$(mktemp -d -t "${wtpfx}-$(basename "$1").XXXXXX")"
        git worktree add -f -q "${_tmpdir}" HEAD
        cd "${_tmpdir}"
                                                                                                                                                                         
        "$@"
    )
}

function verify_generated() {
    run_in_worktree "verify_generated" "$@"
}
```

Then the hack/verify-licenses.sh script can call `run_in_worktree` against a function.

thockin commented on cmd/gotemplate/OWNERS:
```diff
@@ -0,0 +1,11 @@
+  - dims
+  - pohly
+approvers:
+  - dims
```
why not you?

pohly replied:
I've added myself.

thockin commented on cmd/gotemplate/OWNERS:
```diff
@@ -0,0 +1,11 @@
+  - dims
+labels:
+  - sig/testing
+  - sig/contributor-experience
```
I don't think contribex wants to care about this

pohly replied:
I've removed `sig/contributor-experience`.

What do we do about approvers? Just @dims?

thockin commented on cmd/gotemplate/gotemplate.go:
```diff
@@ -0,0 +1,97 @@
+
+func main() {
+	kvs := make(map[string]string)
+	flag.Parse()
```
do we need to call this if we have no flags?

pohly replied:
Not really. It adds `-h`, but that doesn't print anything useful. I've removed `flag`.

thockin commented on hack/update-golangci-lint-config.sh:
```diff
@@ -0,0 +1,50 @@
+generate hack/golangci-hints.yaml Hints=1
+
+# Inform about differences.
+echo
```
I'm not sure we need this output every run.  Maybe just print the commands the user can run to see them?

pohly replied:
My intention was to have that output in the Prow job output when someone reviews a PR which changes the template.  Checking out the PR and running diff locally also works. Removed.

thockin reviewed (commented):
I wonder if we should move tools like this under hack/tools or something?  Not a problem for this PR.

pohly commented:
> Then the hack/verify-licenses.sh script can call run_in_worktree against a function.

Is that really worth the churn? As far as I can tell at a glance, the indention of most of `verify-licenses.sh` would change when moving the current top-level code into a function.

pohly commented on hack/lib/verify-generated.sh:
```diff
@@ -0,0 +1,66 @@
+    kube::util::ensure_clean_working_dir
+
+    # This sets up the environment, like GOCACHE, which keeps the worktree cleaner.
+    kube::golang::setup_env
```
Yes, that's safer. Done.

pohly commented on hack/lib/verify-generated.sh:
```diff
@@ -0,0 +1,66 @@
+    # This sets up the environment, like GOCACHE, which keeps the worktree cleaner.
+    kube::golang::setup_env
+
+    _tmpdir="$(kube::realpath "$(mktemp -d -t "verify-generated-$(basename "$1" | sed -e 's/[^a-zA-Z][^a-zA-Z]*/-/g').XXXXXX")")"
```
No particular reason as far as I remember. I just wanted to ensure that the `verify-generated-` prefix is followed by something using the same convention (alpha, hyphen as separator). Probably not worth it or even confusing when trying to map the temp dir back to the original command - I've removed the sed.

pohly commented:
/test pull-kubernetes-verify-strict-lint

pohly commented:
I pushed [my updates](https://github.com/kubernetes/kubernetes/compare/dba7fb1a6e9cfc4fe02127e240e5da364a4f4cf9..d6da30e5afa97a97d170e3607f8f1991c046f16b).

However, I then had to rebase to include some config change from master in the template.

pohly commented:
/test pull-kubernetes-verify-strict-lint

pohly commented on cmd/gotemplate/gotemplate.go:
```diff
@@ -0,0 +1,97 @@
+//
+// Besides the default functions (https://pkg.go.dev/text/template#hdr-Functions),
+// gotemplate also implements:
+// - include <filename>: returns the content of that file as string
```
See https://github.com/kubernetes/kubernetes/pull/119860 for that usage.

thockin commented on hack/lib/verify-generated.sh:
```diff
@@ -0,0 +1,66 @@
+# in the work tree.
+#
+# Example: kube::verify::generated "Mock files are out of date" "Please run 'hack/update-mocks.sh'" hack/update-mocks.sh
+kube::verify::generated() (
```
This is tricky syntax.  It surprised even me.  Can I arm-twist you into something more obvious:

```
kube::verify::generated() {
  ( # a subshell prevents environment changes from leaking out of this function
       local failure_header=$1
       shift

       # etc.
  )
}
```

pohly replied:
Sure, if that's preferred. Defining commands as a single sub-shell is how I learned it, but I guess it's not that common in Kubernetes.

Updated.

thockin reviewed (commented):
LGTM overall

thockin commented:
Thanks!

/lgtm
/approve
/retest
